source/agora/node/Ledger.d

```diff
@@ -98,54 +106,154 @@ public class Ledger
         Called when a new transaction is received.
 
         If the transaction is accepted it will be added to
-        a new block, and the block will be added to the ledger.
+        the transaction pool. If there are enough valid transactions
+        in the pool, a block will be created.
 
         If the transaction is invalid, it's rejected and false is returned.
 
         Params:
             tx = the received transaction
 
         Returns:
-            true if the transaction is valid and was added to a block
+            true if the transaction is valid and was added to the pool
 
     ***************************************************************************/
 
     public bool acceptTransaction (Transaction tx) @safe
     {
-        if (!tx.isValid(&this.findUTXO))
+        if (!this.isValidTransaction(tx))
+        {
+            logInfo("Rejected tx: %s", tx);
             return false;
+        }
 
         this.pool.add(tx);
         if (this.pool.length >= Block.TxsInBlock)
-            this.makeBlock();
+            this.tryCreateBlock();
 
         return true;
     }
 
     /***************************************************************************
 
-        Create a new block out of transactions in the storage.
+        Add a validated block to the Ledger,
+        and add all of its outputs to the UTXO set.
+
+        Params:
+            block = the block to add
+
+    ***************************************************************************/
+
+    private void addValidatedBlock (const ref Block block) nothrow @safe
+    {
+        // add the new UTXOs
+        block.txs.each!(tx => this.utxo_set.updateUtxoCache(tx));
+
+        // remove the TXs from the Pool
+        block.txs.each!(
+            (tx)
+            {
+                try
+                {
```

Comment:
`scope (failure) ...`

Reply:
👋

Reply:
Love it. (the feature)

I replaced it everywhere else but here, missed it.

Reply:
One of Kenji's many stroke of genius